--- a/device.cpp
+++ b/device.cpp
@@ -91,19 +91,29 @@ static VALUE rb_device_equal(VALUE self, VALUE other) {
   if (!rb_obj_is_kind_of(other, rb_class_of(self))) {
     return Qfalse;
   }
   
-  mx::Device* self_device;
-  mx::Device* other_device;
-  
-  Data_Get_Struct(self, mx::Device, self_device);
-  Data_Get_Struct(other, mx::Device, other_device);
-  
-  return (*self_device == *other_device) ? Qtrue : Qfalse;
+  mx::Device* self_device = nullptr;
+  Data_Get_Struct(self, mx::Device, self_device);

+  // 1) If 'other' is an Integer, interpret it as a DeviceType
+  if (FIXNUM_P(other)) {
+    int other_type = NUM2INT(other);
+    mx::Device temp((mx::Device::DeviceType) other_type, 0);
+    return (*self_device == temp) ? Qtrue : Qfalse;
+  }
+
+  // 2) If 'other' is a Device, do a direct Device == Device comparison
+  if (rb_obj_is_kind_of(other, rb_class_of(self))) {
+    mx::Device* other_device = nullptr;
+    Data_Get_Struct(other, mx::Device, other_device);
+    return (*self_device == *other_device) ? Qtrue : Qfalse;
+  }
+
+  // 3) Otherwise, return false
+  return Qfalse;
 }